app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.cpp

```diff
@@ -0,0 +1,126 @@
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        return {};
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        response.Push<u64>(0);
+        return {};
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfAndResetOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleaved(session, request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleaved(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto reset{request.Pop<bool>()};
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        if (reset)
+            ResetContext();
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        response.Push<u64>(0);
+        return {};
+    }
+
+    void IHardwareOpusDecoder::ResetContext() {
+        opus_decoder_ctl(decoderState, OPUS_RESET_STATE);
+    }
+
+    ResultValue<IHardwareOpusDecoder::DecodeResult> IHardwareOpusDecoder::Decode(span<u8> dataIn, span<opus_int16> dataOut) {
+        if (dataIn.size() <= sizeof(OpusDataHeader)) {
+            return result::BadOpusDataHeader;
+        }
+
+        auto frameSize{util::SwapEndianness(dataIn.as<OpusDataHeader>().size)};
+        auto requiredOutSize{frameSize + sizeof(OpusDataHeader)};
+        if (frameSize > decoderOutputBufferSize || dataOut.size() < requiredOutSize) {
+            return result::OpusBufferTooSmall;
+        }
+
+        dataIn = dataIn.subspan(sizeof(OpusDataHeader));
+        int decodedCount{opus_decode(decoderState, dataIn.data(), frameSize, dataOut.data(), decoderOutputBufferSize, false)};
+
+        if (decodedCount < 0) {
+            Result nnResult{};
+            switch (decodedCount) {
+                case OPUS_BAD_ARG:
+                    nnResult = result::OpusBadArgument;
```

Comment:
```suggestion
                    return result::OpusBadArgument;
```

do for others

Reply:
just a heads up, that switch case is currently useless as for every result returned it throws OpusException no matter what

Reply:
I made it actually useful and now it returns result values to guest